### PR TITLE
Feat: add support for json-patch and json-merge-patch

### PR DESCRIPTION
**Pull request by Somefive**
Signed-off-by: Somefive <yd219913@alibaba-inc.com>

### Description of your changes

Support `json-patch` ([RFC 6902](https://datatracker.ietf.org/doc/html/rfc6902)) and `json-merge-patch` ([RFC 7396](https://datatracker.ietf.org/doc/html/rfc7396)) trait for mutating workload. Example:

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: busybox
spec:
  components:
    - name: busybox
      type: webservice
      properties:
        image: busybox
        cmd: ["sleep", "86400"]
        labels:
          pod-label-key: pod-label-value
          to-delete-label-key: to-delete-label-value
      traits:
        # the json merge patch can be used to add, replace and delete fields
        # the following part will
        # 1. add `deploy-label-key` to deployment labels
        # 2. set deployment replicas to 3
        # 3. set `pod-label-key` to `pod-label-modified-value` in pod labels
        # 4. delete `to-delete-label-key` in pod labels
        # 5. reset `containers` for pod
        - type: json-merge-patch
          properties:
            metadata:
              labels:
                deploy-label-key: deploy-label-added-value
            spec:
              replicas: 3
              template:
                metadata:
                  labels:
                    pod-label-key: pod-label-modified-value
                    to-delete-label-key: null
                spec:
                  containers:
                    - name: busybox-new
                      image: busybox:1.34
                      command: ["sleep", "864000"]
```

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: busybox
spec:
  components:
    - name: busybox
      type: webservice
      properties:
        image: busybox
        cmd: ["sleep", "86400"]
        labels:
          pod-label-key: pod-label-value
          to-delete-label-key: to-delete-label-value
      traits:
        # the json patch can be used to add, replace and delete fields
        # the following part will
        # 1. add `deploy-label-key` to deployment labels
        # 2. set deployment replicas to 3
        # 3. set `pod-label-key` to `pod-label-modified-value` in pod labels
        # 4. delete `to-delete-label-key` in pod labels
        # 5. add sidecar container for pod
        - type: json-patch
          properties:
            operations:
              - op: add
                path: "/metadata"
                value:
                  labels:
                    deploy-label-key: deploy-label-added-value
              - op: add
                path: "/spec/replicas"
                value: 3
              - op: replace
                path: "/spec/template/metadata/labels/pod-label-key"
                value: pod-label-modified-value
              - op: remove
                path: "/spec/template/metadata/labels/to-delete-label-key"
              - op: add
                path: "/spec/template/spec/containers/1"
                value:
                  name: busybox-sidecar
                  image: busybox:1.34
                  command: ["sleep", "864000"]
```

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: docs/examples/traits/json-patch/example.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        # the json patch can be used to add, replace and delete fields
+        # the following part will
+        # 1. add `deploy-label-key` to deployment labels
+        # 2. set deployment replicas to 3
+        # 3. set `pod-label-key` to `pod-label-modified-value` in pod labels
+        # 4. delete `to-delete-label-key` in pod labels
+        # 5. add sidecar container for pod
+        - type: json-patch
+          properties:
+            patch:
```
Comment (wonderflow) [inline docs/examples/traits/json-patch/example.yaml:25]:
```suggestion
            operations:
```

will it be better?

**Review by wonderflow — COMMENTED** [review]
let's increase the coverage by adding more tests

File: docs/examples/traits/json-patch/example.yaml (hunk 1 above)
Comment (Somefive) [reply docs/examples/traits/json-patch/example.yaml:25]: Good idea.

**Comment by Somefive** [comment]
> let's increase the coverage by adding more tests

OK, I will add some to increase the coverage. Seems added tests cannot cover most error branches.

**Review by wonderflow — APPROVED** [review]

File: docs/examples/traits/json-patch/example.yaml (hunk 1 above)
Comment (Somefive) [reply docs/examples/traits/json-patch/example.yaml]: Changed to `operations`.

**Review by FogDong — APPROVED** [review]

**Pull request merged**